clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java

```diff
@@ -3405,6 +3406,54 @@ MAX_REQUEST_SIZE, ACKS_ALL, MAX_RETRIES, new SenderMetricsRegistry(new Metrics(t
         assertEquals(1, transactionManager.sequenceNumber(tp1).intValue());
     }
 
+    @Test
+    public void testMakeIllegalTransitionFatal() {
+        doInitTransactions();
+        assertTrue(transactionManager.isTransactional());
+
+        // Step 1: create a transaction.
+        transactionManager.beginTransaction();
+        assertTrue(transactionManager.hasOngoingTransaction());
+
+        // Step 2: abort a transaction (wait for it to complete) and then verify that the transaction manager is
+        // left in the READY state.
+        TransactionalRequestResult abortResult = transactionManager.beginAbort(TransactionManager.InvalidStateDetectionStrategy.FOREGROUND);
+        runUntil(abortResult::isCompleted);
+        abortResult.await();
+        assertTrue(abortResult.isSuccessful());
+        assertFalse(transactionManager.hasOngoingTransaction());
+        assertTrue(transactionManager.isReady());
+
+        // Step 3: create a batch and simulate the Sender handling a failed batch, which would *attempt* to put
+        // the transaction manager in the ABORTABLE_ERROR state. However, that is an illegal state transition, so
+        // verify that it failed and caused the transaction manager to be put in an unrecoverable FATAL_ERROR state.
+        ProducerBatch batch = batchWithValue(tp0, "test");
+        assertThrowsFatalStateException("handleFailedBatch", () -> transactionManager.handleFailedBatch(batch, new NetworkException("Disconnected from node 4"), false));
+        assertTrue(transactionManager.hasFatalError());
+
+        // Step 4: validate that the transactions can't be started, committed
+        assertThrowsFatalStateException("beginTransaction", () -> transactionManager.beginTransaction());
+        assertThrowsFatalStateException("beginAbort", () -> transactionManager.beginAbort(TransactionManager.InvalidStateDetectionStrategy.FOREGROUND));
+        assertThrowsFatalStateException("beginCommit", () -> transactionManager.beginCommit());
+        assertThrowsFatalStateException("maybeAddPartition", () -> transactionManager.maybeAddPartition(tp0));
+        assertThrowsFatalStateException("initializeTransactions", () -> transactionManager.initializeTransactions());
+        assertThrowsFatalStateException("sendOffsetsToTransaction", () -> transactionManager.sendOffsetsToTransaction(Collections.emptyMap(), new ConsumerGroupMetadata("fake-group-id")));
+    }
+
+    private void assertThrowsFatalStateException(String methodName, Runnable operation) {
+        try {
+            operation.run();
+        } catch (KafkaException t) {
```

Comment:
Just to confirm, fatal errors have KafkaException thrown otherwise, we see IllegalState?

Reply:
@jolshan No. Sadly, it's more convoluted than that 😦 

In the existing implementation, upon detection of the invalid transition in `transitionTo`, we throw an `IllegalStateException`.

In this PR, upon detection of the invalid transition in `transitionTo`, we first "poison" the state variables and **then** throw an `IllegalStateException`. Then later calls to `maybeFailWithError()` will throw a `KafkaException` since `hasError()` is `true` and there's no special case handling for when `lastError` is an instance of `IllegalStateException` .

Reply:
@jolshan Based on the API contract for the transactional methods, it seems like I should add a special case to `maybeFailWithError()` for handling when `lastError` is of type `IllegalStateException`. Thoughts?

Reply:
That may be useful for debugging :)

Reply:
I added a check for `IllegalStateException` in `maybeFailWithError`.